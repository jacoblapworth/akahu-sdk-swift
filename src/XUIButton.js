import React, {PropTypes} from 'react';
import Classes from 'xui-css-classes';
import cn from 'classnames';
import XUIIcon from 'xui-icon';
import XUILoader from 'xui-loader';

// general helpers
const keys = Object.keys;
const values = x => keys(x).map(k => x[k]);

const ButtonClasses = Classes.Button;

/**
 * String constants
 *
 * @private
 * @type {Object}
 */
const CONSTANTS = {
	TYPES: {
		BUTTON: 'button',
		LINK: 'link'
	},
	ELEMENT_TYPES: {
		BUTTON: 'button',
		LINK: 'a'
	},
	BUTTON_TYPES: {
		SUBMIT: 'submit',
		BUTTON: 'button',
		RESET: 'reset'
	},
	VARIANTS: {
		'primary': ButtonClasses.MAIN,
		'create': ButtonClasses.CREATE,
		'link': ButtonClasses.LINK,
		'negative': ButtonClasses.NEGATIVE,
		'standard': ButtonClasses.STANDARD,
		'unstyled': ''
	},
	SIZES: {
		'small': ButtonClasses.SMALL,
		'full-width': Classes.Utility.FULL_WIDTH,
		'full-width-mobile': ButtonClasses.FULL_WIDTH
	}
};

const propTypes = {

	className: PropTypes.string,
	children: PropTypes.node,
	qaHook: PropTypes.string,

	/** @property {boolean} [isDisabled] Determines if the button is disabled or not. */
	isDisabled: PropTypes.bool,

	/** @property {string} [isExternalLink] If true, sets appropriate `rel` values to prevent new page from having access to `window.opener`. Should be used for links pointing at external sites. **/
	isExternalLink: PropTypes.bool,

	/** @property {boolean} [isLoading] If true, shows a loader inside the button and also disables the button to prevent clicking. Can be used in conjunction with isDisabled (which also provides a disabled class)  */
	isLoading: PropTypes.bool,

	/** @property {boolean} [isGrouped] If this button is part of a parent button group */
	isGrouped: PropTypes.bool,

	/** @property {function} onClick Bind a function to fire when the button is clicked */
	onClick: function (props) {
		if (props.type === 'button' && !(typeof props.onClick === 'function')) {
			throw new Error('Non-link buttons require an onClick function.');
		}
	},

	/** @property {function} [onSecondaryClick] Bind a function to fire when the second button in a split button is clicked */
	onSecondaryClick: function (props) {
		if (props.split && !(typeof props.onSecondaryClick === 'function')) {
			throw new Error('Split buttons require a secondary click handler');
		}
	},

	/** @property {string} [variant='standard'] Determines what the purpose of this button is. `standard`, `primary`, `create`, `negative`, `link` or `unstyled`. */
	variant: PropTypes.oneOf(keys(CONSTANTS.VARIANTS)),

	/** @property {string} [size='default'] Modifier for the size of the button. `small`, `full-width`, or `full-width-layout`. */
	size: PropTypes.oneOf(keys(CONSTANTS.SIZES)),

	/** @property {string} [type='button'] The HTML type of this button. `button`, or `link`. Defaults to `button` */
	type: PropTypes.oneOf(values(CONSTANTS.TYPES)),

	/** @property {string} [buttonType='submit'] The type attribute of this button. `submit`, `button`, or `reset`. Defaults to `submit` */
	buttonType: PropTypes.oneOf(values(CONSTANTS.BUTTON_TYPES)),

	/** @property {string} [href] The `href` attribute to use on the anchor element (ignored unless `type` is `link`) */
	href: function (props) {
		if (props.type === CONSTANTS.LINK && !props.onClick && !props.href) {
			throw new Error('Link buttons without an onClick handler require an href.');
		}
	},

	/** @property {string} [rel] The `rel` attribute to use on the anchor element (ignored unless `type` is `link`) */
	rel: PropTypes.string,

	/** @property {number} [tabIndex=0] The HTML tabIndex property to put on the component */
	tabIndex: PropTypes.number,

	/** @property {string} [target] The `target` attribute to use on the anchor element (ignored unless `type` is `link`) */
	target: PropTypes.string,

	/** @property {string} [title] The `title` attribute for this button */
	title: PropTypes.string,

	/** @property {boolean} [split] Changes the button to a split button. Use `onSecondaryClick` with this to create dropdown experiences */
	split: PropTypes.bool
};

/**
 * Default property values for this component.
 *
 * @public
 */
const defaultProps = {
	buttonType: CONSTANTS.BUTTON_TYPES.SUBMIT,
	tabIndex: 0,
	type: CONSTANTS.ELEMENT_TYPES.BUTTON,
	variant: 'standard'
};

/**
 * Returns a class name for the button depending on the button variant string given. Will return
 * undefined if no matching variant is given.
 *
 * @private
 * @param {string} variant - The button variant
 * @return {string} The variant specific classname
 */
const getVariantClass = variant => {
	const variants = CONSTANTS.VARIANTS;
	return variants.hasOwnProperty(variant) ? variants[variant] : ButtonClasses.STANDARD;
};

/**
 * Returns a classname for the button depending on it's disabled state
 *
 * @private
 * @param {boolean} isDisabled - Whether or not the button is disabled
 * @return {string} The disabled state specific classname
 */
const getDisabledClass = isDisabled => isDisabled ? ButtonClasses.IS_DISABLED : null;

/**
 * Returns a class name for the button depending on the button sizing string given. Will return
 * undefined if no matching size is given.
 *
 * @private
 * @param {string} size - The button size
 * @return {string} The size specific class name
 */
const getSizeClass = size => CONSTANTS.SIZES[size];

/**
 * Returns a class name for the button depending on if it has been set to belong to a group
 *
 * @private
 * @param {boolean} isGrouped - Whether or not the button belongs to a group
 * @return {string} The grouped state specific class name
 */
const getGroupClass = isGrouped => isGrouped ? ButtonClasses.GROUPED : null;

/**
 * Replaces any href of `#` or undefined with `javascript:void(0)`. Else returns the passed href.
 *
 * @private
 * @param {string} href - A given link's href
 * @return {string} The href that will be assigned to a link
 */
const getHref = href => (!href || href === '#') ? 'javascript:void(0)' : href;

/**
 * KeyPress handler which will dispatch a click event when the space bar is pressed.
 *
 * @private
 * @param {KeyboardEvent} event
 * @param {Object} props
 */
function handleSpacebarAsClick(event, props) {
	if (props.isDisabled || props.isLoading) {
		let shouldClick;
		if (event.key) {
			shouldClick = event.key === ' ' || event.key === 'Spacebar';
		} else {
			shouldClick = (event.keyCode || event.which) === 32;
		}

		if (shouldClick) {
			// Clicking the space bar causes scrolling by default.  No bueno for a button.
			event.preventDefault();

			// A native event needs to be dispatched to ensure that all
			// browsers will follow the link.
			// Generate a click event with the latest API, if possible.
			// Use document.createEvent for IE 11.
			let clickEvent;
			if (typeof window.Event === 'function') {
				clickEvent = new MouseEvent('click', {
					bubbles: true,
					cancelable: true,
					view: window,
					detail: 0,
					screenX: 0,
					screenY: 0,
					clientX: 0,
					clientY: 0,
					ctrlKey: false,
					altKey: false,
					shiftKey: false,
					metaKey: false,
					button: 0,
					relatedTarget: null
				});
			} else {
				clickEvent = document.createEvent('MouseEvents');
				// If you're seeing a line through initMouseEvent because WebStorm, read the comment above the if statement
				clickEvent.initMouseEvent('click', true, true, 'window', 0, 0, 0, 0, 0, false, false, false, false, 0, null);
			}
			event.target.dispatchEvent(clickEvent);
		}
	}
}

const setupLinkProps = (props, elementProps) => {
	elementProps.role = 'button';
	elementProps.onKeyPress = function(e) {
		handleSpacebarAsClick(e, props);
	};
	elementProps.href = getHref(props.href);
	elementProps.target = props.target;
	elementProps.rel = props.rel;

	if(props.isExternalLink) {
		elementProps.rel = (elementProps.rel ? elementProps.rel + ' ' : '') + 'external noopener noreferrer'
	}

	if (props.isDisabled || props.isLoading) {
		elementProps['aria-disabled'] = true;
	}
};

export default class XUIButton extends React.Component {
	render () {
		const _this = this;
		const props = _this.props;
		const isLink = props.type === CONSTANTS.TYPES.LINK;
		const ElementType = isLink ? CONSTANTS.ELEMENT_TYPES.LINK : CONSTANTS.ELEMENT_TYPES.BUTTON;
		const variantClass = getVariantClass(props.variant);
		const isSplit = props.split && props.onSecondaryClick;
		const isDisabled = props.isDisabled || props.isLoading;
		const children = props.isLoading ? <XUILoader size="small" defaultLayout={false} className={ButtonClasses.LOADER} /> : props.children;

		const classNames = cn(
			ButtonClasses.BASE,
			props.className,
			variantClass,
			getDisabledClass(props.isDisabled),
			getSizeClass(props.size),
			getGroupClass(props.isGrouped)
		);

		const clickHandler = function() {
			if (isLink && isDisabled) {
				event.preventDefault();
			} else if (!isLink || isLink && props.onClick){
				props.onClick.call(_this, ...arguments);
			}
		};

		const secondaryClickHandler = function() {
			if (isLink && isDisabled) {
				event.preventDefault();
			} else {
				props.onSecondaryClick.call(_this, ...arguments);
			}
		};

		// Standard props for all element types
		const elementProps = {
			title: props.title,
			onClick: clickHandler,
			disabled: isDisabled,
			className: classNames,
			tabIndex: isDisabled ? -1 : props.tabIndex
		};

		// Element type specific props
		if (isLink) {
			setupLinkProps(props, elementProps);
		} else {
			elementProps.type = props.buttonType;
		}

		if (isSplit) {
			elementProps.className = `${elementProps.className} ${ButtonClasses.GROUPED}`;
		}

		let Button = (
			<ElementType {...elementProps} data-automationid={props.qaHook}>
				{children}
			</ElementType>
		);

		if (isSplit) {

			Button = (
				<div className={ButtonClasses.GROUPED}>
					{Button}
					<ElementType className={cn(elementProps.className, ButtonClasses.SPLIT)} onClick={secondaryClickHandler}>
						<XUIIcon icon="caret" className={ButtonClasses.CARET}/>
					</ElementType>
				</div>
			);
		}

		return Button;
	}
}

XUIButton.propTypes = propTypes;
XUIButton.defaultProps = defaultProps;
