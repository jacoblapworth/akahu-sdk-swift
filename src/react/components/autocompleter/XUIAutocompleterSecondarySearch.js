import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import XUIAutocompleterTextInput from './XUIAutocompleterTextInput';
import DropDown from '../dropdown/DropDown';
import DropDownToggled from '../dropdown/DropDownToggled';
import searchPath from '@xero/xui-icon/icons/search'
import XUITextInputSideElement from '../textInput/XUITextInputSideElement';
import XUIIcon from '../icon/XUIIcon';
import {ns} from "../helpers/xuiClassNamespace";

import { intervalRunner, isVisible } from './private/helpers';

export default class XUIAutocompleterSecondarySearch extends PureComponent {
	constructor(props) {
		super(props);
		this.openDropDown = this.openDropDown.bind(this);
		this.closeDropDown = this.closeDropDown.bind(this);
		this.onOpen = this.onOpen.bind(this);
		this.highlightItem = this.highlightItem.bind(this);
	}

	/**
	 * Set the state as not hidden in order to toggle the list open.
	 *
	 * @public
	 */
	openDropDown() {
		this.ddt.openDropDown();
	}

	/**
	 * Set the state as hidden in order to toggle the list closed.
	 *
	 * @public
	 */
	closeDropDown() {
		this.ddt.closeDropDown();
	}

	/**
	 * Manually highlight an item in the list for selection.
	 *
	 * @public
	 */
	highlightItem(item) {
		this.dropdown.highlightItem(item);
	}

	/**
	 * Focuses the autocompleter input before calling props.onOpen
	 *
	 * @public
	 */
	onOpen() {
		this.focusInput();
		this.props.onOpen && this.props.onOpen();
	}

	/**
	 * Focus the XUIAutocompleterInput input element, if visible.
	 *
	 * @public
	 */
	focusInput() {
		const inputDOM = this.input;
		const isInputRendered = () => isVisible(inputDOM);
		const setter = () => {
			inputDOM.focus();
		};
		intervalRunner(isInputRendered, setter);
	}

	render() {
		const completer = this;
		const props = completer.props;

		let listQaHook = null;
		let containerQaHook = null;
		if (props.qaHook) {
			listQaHook = `${props.qaHook}--list`;
			containerQaHook = `${props.qaHook}--container`;
		}
		const dropdownClasses = cn(
			props.dropdownSize ? '' : `${ns}-u-fullwidth`,
			props.dropdownClassName,
		);
		const searchItem = (
			<div className={`${ns}-dropdown--header-container`}>
					<XUIAutocompleterTextInput
						value={props.searchValue}
						leftElement={
							<XUITextInputSideElement>
								<XUIIcon path={searchPath} />
							</XUITextInputSideElement>
						}
						placeholder={props.placeholder}
						throttleInterval={props.searchThrottleInterval}
						onChange={props.onSearch}
						inputRef={c => completer.input = c}
						textInputComponentProps={{
							isBorderlessSolid: true
						}}
						inputProps={{
							id: props.inputId
						}}
					/>
			</div>
		);
		const dropdownToggledClasses = props.dropdownSize ? '' : `${ns}-u-fullwidth`;
		const dropdown = (
			<DropDown
				ref={d => completer.dropdown = d}
				ignoreKeyboardEvents={[32,37,39]} /* Space doesn't select in an autocompleter; left and right arrow keys should move cursor in the input */
				hasKeyboardEvents={false}
				onSelect={props.onOptionSelect}
				id={props.id}
				className={dropdownClasses}
				qaHook={listQaHook}
				size={props.dropdownSize}
				fixedWidth={props.dropdownFixedWidth}
				header={searchItem}
				footer={props.footer}
				restrictFocus={props.restrictFocus}
			>
			{props.children}
		</DropDown>);

		return (
			<div
				ref={c => completer.rootNode = c}
				className={props.className}
				data-automationid={containerQaHook}
			>
				<DropDownToggled
					ref={c => completer.ddt = c}
					trigger={props.trigger}
					dropdown={dropdown}
					onOpen={completer.onOpen}
					onClose={props.onClose}
					closeOnSelect={props.closeOnSelect}
					className={dropdownToggledClasses}
					matchTriggerWidth={props.matchTriggerWidth}
				/>
			</div>
		);
	}
}

XUIAutocompleterSecondarySearch.propTypes = {
	/** Callback to handle when an option has been selected from the dropdown */
	onOptionSelect: PropTypes.func,

	/**
	 * When set to true a loader will be displayed instead of the picklist items.
	 * State for this should be managed externally and it's defaulted to false.
	 */
	loading: PropTypes.bool,

	/** ID to be added to the dropdown element of the completer */
	id: PropTypes.string,

	/** Value that should be inside the input. */
	searchValue: PropTypes.string,

	/** CSS class(es) to go on the wrapping DOM node */
	className: PropTypes.string,

	/** CSS class(es) to go on the dropdown list */
	dropdownClassName: PropTypes.string,

	/** CSS class(es) to go on the input */
	inputClassName: PropTypes.string,

	/** Placeholder for the input */
	placeholder: PropTypes.string,

	/** A set of pills to show above the input.  Useful for showing what was selected in a multi-select */
	pills: PropTypes.oneOfType([PropTypes.element, PropTypes.arrayOf(PropTypes.element)]),

	/** Callback for when the list opens */
	onOpen: PropTypes.func,

	/** Callback for when the list closes */
	onClose: PropTypes.func,

	/** Callback for when the user types into the search box */
	onSearch: PropTypes.func,

	/** If you want to throttle the input's onChange handler, put the throttle interval here */
	searchThrottleInterval: PropTypes.number,

	/** Maps to the 'size' property of the dropdown component. */
	dropdownSize: PropTypes.oneOf(['small', 'medium', 'large', 'xlarge']),

	/** Maps to the `closeOnSelect` property of the DropDownToggled component. */
	closeOnSelect: PropTypes.bool,

	/** When set to true the dropdown will automatically open when the input is given focus. */
	openOnFocus: PropTypes.bool,

	/** Will be passed directly down to the DropDownToggled component as the main trigger. */
	trigger: PropTypes.element.isRequired,

	/** ID to be applied to the XUIAutocompleterInput component. */
	inputId: PropTypes.string,

	/** Force the desktop user experience, even if the viewport is narrow enough for mobile. */
	forceDesktop: PropTypes.bool,

	/** If a size is set, this will force the dropdown to that size instead of setting it as a max width. */
	dropdownFixedWidth: PropTypes.bool,

	/** Whether focus should be restricted to the dropdown while it's open. */
	restrictFocus: PropTypes.bool,

	/**
	 * Setting to true will allow the dropdown's width to be set dependent of the trigger width.
	 * Note: Setting this to true will override any size prop on DropDown. XUI design has also decided
	 * to keep a minimum width on the dropdown, so dropdown may not match the width of narrow triggers.
	 */
	matchTriggerWidth: PropTypes.bool,

	/** A Footer element can be added. */
	footer: PropTypes.element,

	qaHook: PropTypes.string,
	children: PropTypes.node,
};

XUIAutocompleterSecondarySearch.defaultProps = {
	loading: false,
	searchThrottleInterval: 0,
	openOnFocus: false,
	inputId: 'secondary_search_input',
	restrictFocus: true,
};
