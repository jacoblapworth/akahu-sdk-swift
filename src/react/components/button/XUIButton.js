import React from 'react';
import PropTypes from "prop-types";
import cn from 'classnames';
import XUILoader from '../loader/XUILoader';
import { VariantClassNames, SizeClassNames, ButtonTypes } from './private/constants';

/**
 * Returns true if the button is a borderless variant
 *
 * @private
 * @param {string} variant - The button variant
 * @return {boolean} True if is a borderless button
 */
const isBorderlessVariant = variant => {
	return variant.indexOf('borderless') > -1;
}

/**
 * Returns true if the button is an icon variant
 *
 * @private
 * @param {string} variant - The button variant
 * @return {boolean} True if is an icon button
 */
const isIconVariant = variant => {
	return variant.indexOf('icon') > -1;
}

/**
 * Returns a class name for the button depending on the button variant string given. Will return
 * undefined if no matching variant is given.
 *
 * @private
 * @param {string} variant - The button variant
 * @return {string} The variant specific class name
 */
const getVariantClass = variant => {
	return VariantClassNames.hasOwnProperty(variant) ? VariantClassNames[variant] : 'xui-button-standard';
};

/**
 * Replaces any href of `#` or undefined with `javascript:void(0)`. Else returns the passed href.
 *
 * @private
 * @param {string} href - A given link's href
 * @return {string} The href that will be assigned to a link
 */
const getHref = href => (!href || href === '#') ? 'javascript:void(0)' : href;

/**
 * KeyPress handler which will dispatch a click event when the space bar is pressed.
 *
 * @private
 * @param {KeyboardEvent} event
 * @param {Object} props
 */
function handleSpacebarAsClick(event, props) {
	if (props.isDisabled || props.isLoading) {
		let shouldClick;
		if (event.key) {
			shouldClick = event.key === ' ' || event.key === 'Spacebar';
		} else {
			shouldClick = (event.keyCode || event.which) === 32;
		}

		if (shouldClick) {
			// Clicking the space bar causes scrolling by default.  No bueno for a button.
			event.preventDefault();

			// A native event needs to be dispatched to ensure that all
			// browsers will follow the link.
			// Generate a click event with the latest API, if possible.
			// Use document.createEvent for IE 11.
			let clickEvent;
			if (typeof window.Event === 'function') {
				clickEvent = new MouseEvent('click', {
					bubbles: true,
					cancelable: true,
					view: window,
					detail: 0,
					screenX: 0,
					screenY: 0,
					clientX: 0,
					clientY: 0,
					ctrlKey: false,
					altKey: false,
					shiftKey: false,
					metaKey: false,
					button: 0,
					relatedTarget: null
				});
			} else {
				clickEvent = document.createEvent('MouseEvents');
				// If you're seeing a line through initMouseEvent because WebStorm, read the comment above the if statement
				clickEvent.initMouseEvent('click', true, true, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null);
			}
			event.target.dispatchEvent(clickEvent);
		}
	}
}

/**
 * Links styled as buttons require a specific set of attributes to ensure proper functionality and accessibility.
 * This function ensures that those props will be added to the element.
 *
 * @private
 * @param {Object} props
 * @param {Object} elementProps
 */
const setupLinkProps = (props, elementProps) => {
	// If this is just a plain link styled to be a button, the button role is not needed.  However, if this is a link
	// which is styled to look AND function like a button, then we'll need the role.
	if (!props.href || props.onClick) {
		elementProps.role = 'button';
	}
	elementProps.onKeyPress = function(e) {
		handleSpacebarAsClick(e, props);
	};
	elementProps.href = getHref(props.href);
	elementProps.target = props.target;
	elementProps.rel = props.rel;

	if (props.isExternalLink) {
		elementProps.rel = (elementProps.rel ? elementProps.rel + ' ' : '') + 'external noopener noreferrer'
	}

	if (props.isDisabled || props.isLoading) {
		elementProps['aria-disabled'] = true;
	}
};

/**
 * Attempt to focus the root DOM node of the given component, if the DOM node exists
 *
 * @private
 * @param {XUIButton} button
 */
const focusRootNode = button => button.rootNode != null && button.rootNode.focus();

export default class XUIButton extends React.Component {
	focus() {
		focusRootNode(this);
		// Apparently there are times when calling focus won't actually do it.  I think
		// React's getting in the way, but I'm not sure yet....
		if (this.rootNode !== document.activeElement) {
			setTimeout(focusRootNode, 0, this);
		}
	}

	hasFocus() {
		return this.rootNode == null ? false : this.rootNode.contains(document.activeElement);
	}

	render () {
		const xuiButton = this;
		const {
			type,
			variant,
			onClick,
			isDisabled,
			isLoading,
			children,
			className,
			onKeyDown,
			tabIndex,
			size,
			isGrouped,
			qaHook,
			isExternalLink,
			href,
			target,
			rel,
			isLink,
			isInverted,
			retainLayout,
			minLoaderWidth,
			...spreadProps
		} = xuiButton.props;
		const ElementType = isLink ? 'a' : 'button';
		const variantClass = getVariantClass(variant);
		const buttonDisabled = isDisabled || isLoading;
		let buttonChildren = children;

		const loader = (
			<XUILoader
				key={retainLayout && isLoading ? 'button-loader' : null}
				retainLayout={retainLayout}
				size="small"
				defaultLayout={false}
				className="xui-button--loader" />
		);

		if (retainLayout && isLoading) {
			buttonChildren = [
				<div className="xui-u-hidden-content" key='button-children'>{children}</div>,
				loader
			]
		} else if (isLoading) {
			buttonChildren = loader;
		}

		const buttonClassNames = cn('xui-button', className, variantClass, SizeClassNames[size], {
			'xui-button-is-disabled': isDisabled,
			'xui-button-grouped': isGrouped,
			'xui-button-inverted': isInverted && !isBorderlessVariant(variantClass) && !isIconVariant(variantClass),
			'xui-button-borderless-inverted': isInverted && isBorderlessVariant(variantClass) && !isIconVariant(variantClass),
			'xui-button-min-loader-width': minLoaderWidth
		});

		const clickHandler = function() {
			if (isLink && buttonDisabled) {
				event.preventDefault();
			} else if (onClick) {
				onClick.call(xuiButton, ...arguments);
			}
		};

		// Standard props for all element types
		const elementProps = {
			...spreadProps,
			onClick: clickHandler,
			onKeyDown: buttonDisabled ? null : onKeyDown,
			disabled: buttonDisabled,
			className: buttonClassNames,
			tabIndex: buttonDisabled ? -1 : tabIndex
		};

		// Element type specific props
		if (isLink) {
			setupLinkProps({ href, onClick, isExternalLink, target, rel, isDisabled, isLoading }, elementProps);
		} else {
			elementProps.type = type;
		}

		return (
			<ElementType ref={n => xuiButton.rootNode = n} {...elementProps} data-automationid={qaHook}>
				{buttonChildren}
			</ElementType>
		);
	}
}

XUIButton.propTypes = {
	className: PropTypes.string,
	children: PropTypes.node,
	qaHook: PropTypes.string,

	/** Determines if the button is disabled or not. */
	isDisabled: PropTypes.bool,

	/** If true, sets appropriate `rel` values to prevent new page from having access to `window.opener`. Should be used for links pointing at external sites. */
	isExternalLink: PropTypes.bool,

	/** If true, shows a loader inside the button and also disables the button to prevent clicking. Can be used in conjunction with isDisabled (which also provides a disabled class)  */
	isLoading: PropTypes.bool,

	/** If this button is part of a parent button group */
	isGrouped: PropTypes.bool,

	/** A keydown event handler for the button */
	onKeyDown: PropTypes.func,

	/** Bind a function to fire when the button is clicked */
	onClick: PropTypes.func,

	/** Determines the styling variation to apply: `standard`, `primary`, `create`, `negative`, `link`, 'borderless-standard', 'borderless-primary', 'borderless-create', 'borderless-negative', 'borderless-negative', 'icon', 'icon-large', 'icon-inverted', 'icon-inverted-large' or `unstyled`. */
	variant: PropTypes.oneOf(Object.keys(VariantClassNames)),

	/** Modifier for the size of the button. `small`, `full-width`, or `full-width-layout`. */
	size: PropTypes.oneOf(Object.keys(SizeClassNames)),

	/** Whether or not to render this button using an <a> tag */
	isLink: PropTypes.bool,

	/** The type attribute of this button. `submit`, `button`, or `reset`. */
	type: PropTypes.oneOf(Object.keys(ButtonTypes).map(type => ButtonTypes[type])), // Can't use Object.values without polyfilling for older supported browsers

	/** The `href` attribute to use on the anchor element (ignored unless `isLink` is `true`) */
	href: PropTypes.string,

	/** The `rel` attribute to use on the anchor element (ignored unless `isLink` is `true`) */
	rel: PropTypes.string,

	/** The HTML tabIndex attribute value */
	tabIndex: PropTypes.number,

	/** The `target` attribute to use on the anchor element (ignored unless `isLink` is `true`) */
	target: PropTypes.string,

	/** The `title` attribute */
	title: PropTypes.string,

	/** Applies inverted styling */
	isInverted: PropTypes.bool,

	/** When used with `isLoading` this allows the button to retain children width */
	retainLayout: PropTypes.bool,

	/** Use this to specify a min width on the button, when you want to swap to loading states */
	minLoaderWidth: PropTypes.bool
};

XUIButton.defaultProps = {
	tabIndex: 0,
	type: ButtonTypes.button,
	variant: 'standard',
	isLink: false,
	isDisabled: false,
	isExternalLink: false,
	isGrouped: false,
	isLoading: false,
	retainLayout: true
};
