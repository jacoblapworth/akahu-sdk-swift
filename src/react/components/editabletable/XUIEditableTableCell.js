import React from 'react';
import PropTypes from 'prop-types';

import cn from 'classnames';

import { tableName } from './private/constants';

const baseName = `${tableName}cell`;

const XUIEditableTableCell = ({ children, className, width, ...spreadProps }) => {
  return (
    <td className={cn(baseName, className)} style={{ width }} {...spreadProps}>
      {children}
    </td>
  );
};

XUIEditableTableCell.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  width: PropTypes.string,
};

export default XUIEditableTableCell;
