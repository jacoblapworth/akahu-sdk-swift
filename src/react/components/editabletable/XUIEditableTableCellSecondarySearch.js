import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import XUIEditableTableCell from './XUIEditableTableCell';
import XUIAutocompleterSecondarySearch from '../autocompleter/XUIAutocompleterSecondarySearch';
import { tableName } from './private/constants';
import { fixedWidthDropdownSizes } from '../dropdown/private/constants';

const baseName = `${tableName}cellsecondarysearch`;

const XUIEditableTableCellSecondarySearch = ({ cellProps = {}, className, ...spreadProps }) => {
  return (
    <XUIEditableTableCell {...cellProps} className={cn(baseName, cellProps.className)}>
      <XUIAutocompleterSecondarySearch
        {...spreadProps}
        className={cn(`${baseName}--control`, className)}
        isInputLabelHidden
      />
    </XUIEditableTableCell>
  );
};

XUIEditableTableCellSecondarySearch.propTypes = {
  cellProps: PropTypes.object,
  className: PropTypes.string,
  qaHook: PropTypes.string,
  /** Callback to handle when an option has been selected from the dropdown */
  onOptionSelect: PropTypes.func,

  /** ID to be added to the dropdown element of the completer */
  id: PropTypes.string,

  /** Value that should be inside the input. */
  searchValue: PropTypes.string,

  /** CSS class(es) to go on the dropdown list */
  dropdownClassName: PropTypes.string,

  /** CSS class(es) to go on the input */
  inputClassName: PropTypes.string,

  /** CSS class(es) to go on the input container */
  inputContainerClassName: PropTypes.string,

  /** Attributes to set on the native input element */
  inputProps: PropTypes.object,

  /** Placeholder for the input */
  placeholder: PropTypes.string,

  /** Callback for when the list opens */
  onOpen: PropTypes.func,

  /** Callback for when the list closes */
  onClose: PropTypes.func,

  /** Callback for when the user types into the search box */
  onSearch: PropTypes.func.isRequired,

  /** If you want to throttle the input's onChange handler, put the throttle interval here */
  searchThrottleInterval: PropTypes.number,

  /** Maps to the 'size' property of the dropdown component. */
  dropdownSize: PropTypes.oneOf(Object.keys(fixedWidthDropdownSizes)),

  /** Maps to the `closeOnSelect` property of the DropdownToggled component. */
  closeOnSelect: PropTypes.bool,

  /** Maps to the `closeOnTab` property of the DropdownToggled component. Set to false,
   * if you've supplied a footer element with any links or interaction. */
  closeOnTab: PropTypes.bool,

  /** Callback for adding additional onKeyPress functionality */
  onKeyDown: PropTypes.func,

  /** Will be passed directly down to the DropdownToggled component as the main trigger. */
  trigger: PropTypes.element.isRequired,

  /** ID to be applied to the input element. */
  inputId: PropTypes.string,

  /** Label to show above the input */
  inputLabel: PropTypes.node,

  /** Whether to allow the dropdown to take the full width of the wrapper (as SelectBox)
   * or wrap with an inline block. */
  isBlock: PropTypes.bool,

  /** Should label be applied as an aria-label, rather than being visibly displayed. */
  isInputLabelHidden: PropTypes.bool,

  /** Force the desktop user experience, even if the viewport is narrow enough for mobile. */
  forceDesktop: PropTypes.bool,

  /** If a size is set, this will force the dropdown to that size instead of setting it
   * as a max width. */
  dropdownFixedWidth: PropTypes.bool,

  /** Whether focus should be restricted to the dropdown while it's open. */
  restrictFocus: PropTypes.bool,

  /**
   * Setting to true will allow the dropdown's width to be set dependent of the trigger width. <br>
   * **Note:** *Setting this to true will override any size prop on Dropdown.* <br>
   * XUI design has also decided to keep a minimum width on the dropdown,
   * so dropdown may not match the width of narrow triggers. Defaults to true.
   */
  matchTriggerWidth: PropTypes.bool,

  /** A Footer element can be added. */
  footer: PropTypes.element,
};

export default XUIEditableTableCellSecondarySearch;

XUIEditableTableCellSecondarySearch.defaultProps = {
  matchTriggerWidth: true,
};
