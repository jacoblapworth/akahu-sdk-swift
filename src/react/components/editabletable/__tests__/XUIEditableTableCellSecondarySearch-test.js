import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

import XUIButton from '../../button/XUIButton';
import XUIAutocompleterSecondarySearch from '../../autocompleter/XUIAutocompleterSecondarySearch';
import XUIEditableTableCell from '../XUIEditableTableCell';
import XUIEditableTableCellSecondarySearch from '../XUIEditableTableCellSecondarySearch';

Enzyme.configure({ adapter: new Adapter() });

describe('<XUIEditableTableCellSecondarySearch />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <XUIEditableTableCellSecondarySearch
        onSearch={() => {}}
        trigger={<XUIButton>Click me</XUIButton>}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('spreads cellProps onto the table cell', () => {
    const wrapper = mount(
      <table>
        <tbody>
          <tr>
            <XUIEditableTableCellSecondarySearch
              cellProps={{ width: '100px' }}
              onSearch={() => {}}
              trigger={<XUIButton>Click me</XUIButton>}
            />
          </tr>
        </tbody>
      </table>,
    );
    expect(
      wrapper
        .find(XUIEditableTableCell)
        .getDOMNode()
        .getAttribute('width'),
    ).toBe('100px');
  });

  it('spreads the rest of the props onto XUIAutocompleterSecondarySearch', () => {
    const spreadProps = {
      id: 'testId',
      onSearch: () => {},
      trigger: <XUIButton>Click me</XUIButton>,
    };

    const wrapper = mount(
      <table>
        <tbody>
          <tr>
            <XUIEditableTableCellSecondarySearch {...spreadProps} />
          </tr>
        </tbody>
      </table>,
    );
    expect(wrapper.find(XUIAutocompleterSecondarySearch).props()).toMatchObject(spreadProps);
  });
});
