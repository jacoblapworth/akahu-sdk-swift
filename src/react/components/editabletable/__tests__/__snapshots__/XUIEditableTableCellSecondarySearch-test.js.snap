// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<XUIEditableTableCellSecondarySearch /> renders correctly 1`] = `
<XUIEditableTableCell
  className="xui-editabletablecellsecondarysearch"
>
  <XUIAutocompleterSecondarySearch
    className="xui-editabletablecellsecondarysearch--control"
    isInputLabelHidden={true}
    matchTriggerWidth={true}
    onSearch={[Function]}
    restrictFocus={true}
    searchThrottleInterval={0}
    trigger={
      <XUIButton
        fullWidth="never"
        isDisabled={false}
        isExternalLink={false}
        isGrouped={false}
        isLink={false}
        isLoading={false}
        retainLayout={true}
        tabIndex={0}
        type="button"
        variant="standard"
      >
        Click me
      </XUIButton>
    }
  />
</XUIEditableTableCell>
`;
